File: crates/core/src/facade/git.rs
//! The version-control commands and queries adapters call (context C8): what git says about a
//! project, and what a surface may change about it.
//!
//! Reads are ungated. Changes are gated in the core on the user having trusted the project to
//! be changed — a general authorisation kept per project, which version control's write side is
//! the first thing to spend. Every change announces itself the same way the watcher does, by
//! re-reading the status and publishing only if it turned out different, so an action and the
//! watcher noticing the same action converge on one snapshot instead of racing to two.

use std::path::{Path, PathBuf};

use super::Facade;
use crate::events::DomainEvent;
use crate::git::{DiffExtent, Git, GitError, GitStatus, GitWriteError, Prompting};
use crate::ids::ProjectId;
use crate::ports::StoreError;
use crate::vcs::{Branches, DiffTarget, FileContent, FileDiff, HunkRange, ProjectFile};

/// What every exchange this façade starts is allowed to do about a credential only a person can
/// give: ask them. The local user clicked something and is watching, so a dialog is the right
/// answer for them — the opposite decision, for a caller nobody is watching, is named once in
/// [`scoped_git`](super::scoped_git).
const AT_THE_WINDOW: Prompting = Prompting::Allowed;

impl Facade {
    /// A project's working-tree status: what is checked out, how it stands against its
    /// upstream, and every path that differs from the last commit. The snapshot half of
    /// snapshot-then-deltas for version control — pair it with
    /// [`DomainEvent::GitStatusChanged`](crate::events::DomainEvent::GitStatusChanged).
    ///
    /// `Ok(None)` for a project that is not a repository: not keeping a folder under version
    /// control is an ordinary choice, so a surface shows nothing rather than an error. The
    /// remembered status is served when there is one, so several surfaces asking at once cost
    /// one read.
    ///
    /// Reading a repository runs an external tool, so callers reach this through
    /// [`Facade::blocking`] rather than a runtime worker.
    pub fn git_status(&self, project: ProjectId) -> Result<Option<GitStatus>, GitReadError> {
        let root = self.git_root(project)?;
        Ok(self.git.status(project, &root)?)
    }

    /// Every path in `project`'s repository: what it tracks, what it does not yet track, and
    /// what it was told to ignore — the answer a surface browsing the project's files renders.
    /// An ignored directory arrives as itself rather than its contents.
    ///
    /// `Ok(None)` for a project that is not a repository, for the same reason
    /// [`Facade::git_status`] gives it. Reading a repository runs an external tool, so callers
    /// reach this through [`Facade::blocking`] rather than a runtime worker.
    pub fn git_files(&self, project: ProjectId) -> Result<Option<Vec<ProjectFile>>, GitReadError> {
        let root = self.git_root(project)?;
        Ok(self.git.files(project, &root)?)
    }

    /// How one path differs, `target` deciding against what and `extent` how much of the answer
    /// is carried. A diff longer than one capped read says so, and asking again at
    /// [`DiffExtent::Full`] carries the rest.
    ///
    /// `Ok(None)` for a project that is not a repository and for a path that does not name
    /// something inside it. Reading a repository runs an external tool, so callers reach this
    /// through [`Facade::blocking`] rather than a runtime worker.
    pub fn git_diff(
        &self,
        project: ProjectId,
        path: &str,
        target: DiffTarget,
        extent: DiffExtent,
    ) -> Result<Option<FileDiff>, GitReadError> {
        let root = self.git_root(project)?;
        Ok(self.git.diff(project, &root, path, target, extent)?)
    }

    /// The working tree's copy of one path, for a surface showing a file rather than a change
    /// to one. Bounded: a file past the adapter's ceiling arrives only as far as that ceiling
    /// and says so, and one holding bytes that are not text arrives with no text at all.
    ///
    /// `Ok(None)` for a project that is not a repository, for a path outside it, and for one
    /// that is no longer there. Reading a file goes through the repository adapter, so callers
    /// reach this through [`Facade::blocking`] rather than a runtime worker.
    pub fn git_file(
        &self,
        project: ProjectId,
        path: &str,
    ) -> Result<Option<FileContent>, GitReadError> {
        let root = self.git_root(project)?;
        Ok(self.git.file(project, &root, path)?)
    }

    /// The branches `project` could switch to, most recently committed to first and bounded at
    /// [`BRANCH_PAGE_SIZE`](crate::git::BRANCH_PAGE_SIZE), plus whether it has anything stashed —
    /// the two things a branch switcher offers.
    ///
    /// `Ok(None)` for a project that is not a repository, for the same reason
    /// [`Facade::git_status`] gives it. A read, so it is ungated; it runs an external tool, so
    /// callers reach it through [`Facade::blocking`].
    pub fn git_branches(&self, project: ProjectId) -> Result<Option<Branches>, GitReadError> {
        let root = self.git_root(project)?;
        Ok(self.git.branches(project, &root)?)
    }

    /// Whether the user has trusted `project` to be changed by Soloist. A surface asks so it can
    /// offer the trust affordance rather than let an action fail; the gate itself is spent in
    /// the core, so a surface that does not ask changes nothing either.
    pub fn is_project_trusted(&self, project: ProjectId) -> Result<bool, GitWriteError> {
        self.git.is_trusted(project)
    }

    /// Records that trust for `project`, which is what the affordance behind
    /// [`Facade::is_project_trusted`] does. One method behind the gate, so every surface grants
    /// it identically.
    ///
    /// It announces itself on the same event a change to the working tree does: what a
    /// version-control surface may show has changed, which is what that event asks a surface to
    /// go and re-read.
    pub fn trust_project(&self, project: ProjectId) -> Result<(), GitWriteError> {
        self.trust.trust_project(project)?;
        self.bus.publish(DomainEvent::GitStatusChanged { project });
        Ok(())
    }

    /// Hands one of `project`'s files to whatever this machine has registered to open it.
    ///
    /// Gated on trust in the core, because it starts a program the desktop chooses from a name the
    /// repository supplied. A path that leaves the repository — by saying so, or by being a link
    /// that leads out of it — is refused and nothing is opened.
    ///
    /// Runs outside this process, so callers reach this through [`Facade::blocking`] rather than a
    /// runtime worker.
    pub fn git_open_file(&self, project: ProjectId, path: &str) -> Result<(), GitWriteError> {
        let root = self
            .project_root(project)?
            .ok_or(GitWriteError::UnknownProject)?;
        self.git.open_file(project, &root, path)
    }

    /// Records everything the working tree holds for `path` in the index.
    ///
    /// Runs an external tool, so callers reach this through [`Facade::blocking`] rather than a
    /// runtime worker.
    pub fn git_stage(&self, project: ProjectId, path: &str) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| git.stage(project, root, path))
    }

    /// Takes `path` back out of the index, leaving the working tree untouched.
    pub fn git_unstage(&self, project: ProjectId, path: &str) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| git.unstage(project, root, path))
    }

    /// Throws away what the working tree holds for `path` beyond the index. Destructive, and
    /// bounded: it restores from the index, so nothing staged or committed is within its reach.
    /// A path version control does not track is refused rather than deleted.
    pub fn git_discard(&self, project: ProjectId, path: &str) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| git.discard(project, root, path))
    }

    /// Records only one hunk of `path`'s unstaged change in the index. `hunk` names it by where
    /// it falls, so a request built against a diff the file has moved past is refused.
    pub fn git_stage_hunk(
        &self,
        project: ProjectId,
        path: &str,
        hunk: HunkRange,
    ) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| {
            git.stage_hunk(project, root, path, hunk)
        })
    }

    /// Takes only one hunk of `path`'s staged change back out of the index.
    pub fn git_unstage_hunk(
        &self,
        project: ProjectId,
        path: &str,
        hunk: HunkRange,
    ) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| {
            git.unstage_hunk(project, root, path, hunk)
        })
    }

    /// Throws away only one hunk of `path`'s unstaged change. Destructive, and bounded exactly
    /// as [`Facade::git_discard`] is.
    pub fn git_discard_hunk(
        &self,
        project: ProjectId,
        path: &str,
        hunk: HunkRange,
    ) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| {
            git.discard_hunk(project, root, path, hunk)
        })
    }

    /// Records `project`'s index as a commit carrying `message`, or replaces the last commit
    /// with it when `amend`. The user's hooks, signing and configuration all apply, because it
    /// is their own `git` that runs.
    pub fn git_commit(
        &self,
        project: ProjectId,
        message: &str,
        amend: bool,
    ) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| {
            git.commit(project, root, message, amend)
        })
    }

    /// The message a new commit in `project` starts from, as the repository's own configuration
    /// supplies it (`commit.template`), with the guidance lines version control strips from an
    /// edited message already gone. `Ok(None)` where nothing is configured.
    ///
    /// Offered rather than applied: it is the text a message box starts at, and whoever is writing
    /// the commit replaces it. Nothing is committed here, and the commit itself is unchanged —
    /// version control consults a template only when it would open an editor, and Soloist commits
    /// the message it was handed.
    ///
    /// Gated on trust like a change, because the configuration read is one the repository can
    /// carry and it names a file for Soloist to read. Runs an external tool, so callers reach this
    /// through [`Facade::blocking`] rather than a runtime worker.
    pub fn git_commit_template(&self, project: ProjectId) -> Result<Option<String>, GitWriteError> {
        let root = self
            .project_root(project)?
            .ok_or(GitWriteError::UnknownProject)?;
        self.git.commit_template(project, &root)
    }

    /// Starts a branch called `name` at what `project` has checked out, and switches to it.
    pub fn git_create_branch(&self, project: ProjectId, name: &str) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| git.create_branch(project, root, name))
    }

    /// Checks out `project`'s branch called `name`. A switch that would overwrite uncommitted work
    /// is refused, carrying version control's own account of what is in the way — nothing is
    /// stashed or discarded to get past it.
    pub fn git_switch_branch(&self, project: ProjectId, name: &str) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| git.switch_branch(project, root, name))
    }

    /// Removes `project`'s branch called `name`. Destructive, so a surface confirms it first — and
    /// bounded, because a branch holding commits nothing else holds is refused and stays refused.
    pub fn git_delete_branch(&self, project: ProjectId, name: &str) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| git.delete_branch(project, root, name))
    }

    /// Sets what `project`'s working tree holds aside, leaving it as the last commit left it.
    pub fn git_stash(&self, project: ProjectId) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| git.stash(project, root))
    }

    /// Puts `project`'s most recently stashed changes back. A collision with what the working tree
    /// holds now comes back as version control's own account of it, because it left a conflict to
    /// resolve rather than doing what was asked.
    pub fn git_pop_stash(&self, project: ProjectId) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| git.pop_stash(project, root))
    }

    /// Hands the checked-out branch's commits to its remote, publishing the branch when it tracks
    /// nothing yet.
    ///
    /// Reaches a machine Soloist has no say over, under the user's own credentials — Soloist keeps
    /// none of its own and names no credential helper. This is the **local user's** door, so
    /// [`Prompting::Allowed`]: they clicked something and are sitting in front of it, so version
    /// control may ask them for a credential. Bounded by the adapter's own limit for reaching a
    /// remote and stoppable before then ([`Facade::git_stop_exchange`]), so callers reach this
    /// through [`Facade::blocking`].
    pub fn git_push(&self, project: ProjectId) -> Result<(), GitWriteError> {
        self.git_exchange(project, AT_THE_WINDOW, Git::push)
    }

    /// Brings the remote's commits into `project` and reconciles them with what is checked out,
    /// however the user's own configuration says to. Where they have not said, version control
    /// refuses rather than choosing, and its refusal is what comes back.
    pub fn git_pull(&self, project: ProjectId) -> Result<(), GitWriteError> {
        self.git_exchange(project, AT_THE_WINDOW, Git::pull)
    }

    /// Brings the remote's commits in without touching `project`'s working tree, which is what
    /// makes its standing against the upstream true again.
    pub fn git_fetch(&self, project: ProjectId) -> Result<(), GitWriteError> {
        self.git_exchange(project, AT_THE_WINDOW, Git::fetch)
    }

    /// Asks the exchange with a remote running against `project` to stop.
    ///
    /// Instant and infallible: it sets a signal the exchange looks at, taking nothing that exchange
    /// holds — which is what lets it be called while the exchange is still running, from a surface
    /// whose only other option would be to wait the limit out. The exchange itself then ends as
    /// [`GitError::Stopped`](crate::git::GitError::Stopped), the project's gate is released, and the
    /// next read of the repository goes straight through. Nothing to stop is not a failure.
    pub fn git_stop_exchange(&self, project: ProjectId) {
        self.git.stop_exchange(project);
    }

    /// Abandons a merge that is under way in `project`. Destructive within that merge — a conflict
    /// resolved by hand since it began goes with it — so a surface confirms it first.
    pub fn git_abort_merge(&self, project: ProjectId) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| git.abort_merge(project, root))
    }

    /// The one route from a façade to a remote, so which caller may be asked for a credential is
    /// decided in exactly two places — here for the local user, and in the session-scoped surface
    /// for a caller nobody is sitting in front of — rather than at each of the three exchanges.
    ///
    /// `exchange` is one of [`Git::push`], [`Git::pull`] and [`Git::fetch`]; the rest of the
    /// shape (root, gate, announce) is [`Facade::git_change`]'s.
    pub(in crate::facade) fn git_exchange(
        &self,
        project: ProjectId,
        prompting: Prompting,
        exchange: impl FnOnce(&Git, ProjectId, &Path, Prompting) -> Result<(), GitWriteError>,
    ) -> Result<(), GitWriteError> {
        self.git_change(project, |git, root| exchange(git, project, root, prompting))
    }

    /// The shape every version-control change shares: resolve the project's root, make the
    /// change, then re-read the status and announce it if it turned out different.
    fn git_change(
        &self,
        project: ProjectId,
        change: impl FnOnce(&Git, &Path) -> Result<(), GitWriteError>,
    ) -> Result<(), GitWriteError> {
        let root = self
            .project_root(project)?
            .ok_or(GitWriteError::UnknownProject)?;
        change(&self.git, &root)?;
        self.announce_git(project, &root);
        Ok(())
    }

    /// Re-reads `project`'s status and announces it only if it turned out different — the half of
    /// every version-control change that happens after the change itself, so an action and the
    /// watcher noticing that same action converge on one snapshot rather than racing to two.
    pub(super) fn announce_git(&self, project: ProjectId, root: &Path) {
        if matches!(self.git.refresh(project, root), Ok(true)) {
            self.bus.publish(DomainEvent::GitStatusChanged { project });
        }
    }

    /// The folder a project's repository is read from — one resolution behind every read above,
    /// so a project that is not open is refused identically whichever one asked.
    fn git_root(&self, project: ProjectId) -> Result<PathBuf, GitReadError> {
        self.project_root(project)?
            .ok_or(GitReadError::UnknownProject)
    }
}

/// Why a version-control read produced no answer: the project is not open, a durable read
/// failed, or the repository itself could not be read.
#[derive(Debug, thiserror::Error)]
pub enum GitReadError {
    #[error("no such project")]
    UnknownProject,
    #[error(transparent)]
    Store(#[from] StoreError),
    #[error(transparent)]
    Git(#[from] GitError),
}

#[cfg(test)]
#[path = "git_tests.rs"]
mod tests;

File: crates/app/ui/src/lib/git.ts
import type {
  BranchInfo,
  ChangeKind,
  CheckState,
  GitFileStatus,
  HunkRange,
  MergeMethod,
  ReviewThread,
} from "@/domain";

/**
 * The single source for turning a version-control change into its display, alongside
 * `lib/status.ts` for process status. Redundant by the same rule: a letter *and* a colour *and*
 * a word, so a change survives colour blindness and a grayscale screenshot. The exhaustive
 * Record makes the compiler require an entry for every kind the core can report.
 */
export interface ChangeDisplay {
  /** The letter shown beside the path — the same one version control prints. */
  letter: string;
  /** Human label, e.g. "Modified". The letter's accessible name, and its tooltip. */
  label: string;
  /** Tailwind text-color utility, bound to a `--git-*` token. */
  toneClass: string;
  /** A path that no longer exists is drawn struck through, so the row reads as gone. */
  gone: boolean;
}

export const CHANGE: Record<ChangeKind, ChangeDisplay> = {
  modified: { letter: "M", label: "Modified", toneClass: "text-git-modified", gone: false },
  type_changed: {
    letter: "T",
    label: "Type changed",
    toneClass: "text-git-modified",
    gone: false,
  },
  added: { letter: "A", label: "Added", toneClass: "text-git-added", gone: false },
  deleted: { letter: "D", label: "Deleted", toneClass: "text-git-deleted", gone: true },
  renamed: { letter: "R", label: "Renamed", toneClass: "text-git-added", gone: false },
  copied: { letter: "C", label: "Copied", toneClass: "text-git-added", gone: false },
  untracked: { letter: "U", label: "Untracked", toneClass: "text-git-added", gone: false },
  conflicted: { letter: "C", label: "Conflicted", toneClass: "text-git-conflicted", gone: false },
};

/** The tone a path version control was told to ignore is drawn in, on the Files tab. */
export const IGNORED_TONE_CLASS = "text-git-ignored";

/**
 * The one change a row shows for a path changed on both sides of the index. The working tree is
 * what the user is looking at and what they can act on, so it wins; the staged change is the
 * one already recorded. A path with neither is not a change at all.
 */
export function primaryChange(status: GitFileStatus): ChangeKind | null {
  return status.unstaged ?? status.staged;
}

// Which change a folder takes from its children. A folder is a summary, so it shows the child
// that most wants attention: an unresolved conflict first, then a change to what is tracked,
// then a path version control does not track yet. An exhaustive Record rather than a chain of
// comparisons, for the same reason the display map is one — a kind added to the union stops the
// build here until it has been placed.
const SEVERITY: Record<ChangeKind, number> = {
  conflicted: 5,
  deleted: 4,
  modified: 3,
  type_changed: 3,
  renamed: 2,
  copied: 2,
  added: 2,
  untracked: 1,
};

/** The stronger of two changes — what a folder inherits from a child. */
export function strongerChange(a: ChangeKind, b: ChangeKind): ChangeKind {
  return SEVERITY[b] > SEVERITY[a] ? b : a;
}

/**
 * How much of a path is recorded for the next commit, which is what its checkbox shows. A path
 * changed on both sides is neither staged nor unstaged but part of each, and says so rather than
 * rounding to one — ticking it then stages the rest.
 */
export type StagedState = "staged" | "unstaged" | "partial";

export function stagedState(status: GitFileStatus): StagedState {
  if (status.staged === null) return "unstaged";
  return status.unstaged === null ? "staged" : "partial";
}

/**
 * The key one action on one hunk is tracked by while it is in flight. Built from where the hunk
 * falls, which is also how the core names it — so a row that scrolls out of view and back comes
 * back to the same state rather than to whatever the new row in that position holds.
 */
export function hunkKey(path: string, hunk: HunkRange): string {
  return `${path}@${hunk.old_start},${hunk.old_lines},${hunk.new_start},${hunk.new_lines}`;
}

/**
 * The single source for turning a check's state into its display, on the same redundant rule the
 * change map follows — a glyph *and* a tone *and* a word. The exhaustive Record makes the compiler
 * require an entry for every state the core can report.
 */
export interface CheckDisplay {
  /** What the state is called, which is also the glyph's accessible name. */
  label: string;
  /** Tailwind text-color utility, bound to the same `--git-*` tokens the change letters use, so
   *  version control spends one saturated vocabulary rather than two. */
  toneClass: string;
}

export const CHECK: Record<CheckState, CheckDisplay> = {
  pending: { label: "Running", toneClass: "text-git-modified" },
  passed: { label: "Passed", toneClass: "text-git-added" },
  failed: { label: "Failed", toneClass: "text-git-deleted" },
  skipped: { label: "Skipped", toneClass: "text-git-ignored" },
  cancelled: { label: "Cancelled", toneClass: "text-git-ignored" },
  unknown: { label: "Unrecognised", toneClass: "text-git-ignored" },
};

/** How a merge method is named to somebody choosing one. */
export const MERGE_METHOD: Record<MergeMethod, string> = {
  merge: "Create a merge commit",
  squash: "Squash and merge",
  rebase: "Rebase and merge",
};

/** Where a conversation hangs, or null for one about the change as a whole. */
export function threadPlace(thread: ReviewThread): string | null {
  if (thread.path === null) return null;
  return thread.line === null ? thread.path : `${thread.path}:${thread.line}`;
}

/**
 * The single source for turning what is checked out into its display, on the same redundant rule the
 * change map follows: a word beside the tone. Every standing a branch can be in is named, so a
 * branch nobody has published never reads as one that matches its upstream on hue alone.
 */
export interface BranchStanding {
  /**
   * How the branch stands against its upstream, in words. `null` only on a detached head, where
   * there is no branch to have an upstream and the name slot says so itself.
   */
  label: string | null;
  /** Tailwind text-color utility, bound to a `--git-branch-*` token. */
  toneClass: string;
  /** Commits the branch holds that its upstream does not, `0` where there is no comparison. */
  ahead: number;
  /** Commits the upstream holds that the branch does not, `0` where there is no comparison. */
  behind: number;
}

/** The tone a branch known to match its upstream carries. */
const SYNCED_TONE_CLASS = "text-git-branch-synced";
/** The tone every other standing carries: nothing to compare against, or something left to settle. */
const UNSETTLED_TONE_CLASS = "text-git-branch-local";

/** A branch that tracks nothing: it exists here and nowhere else this repository knows of. */
const LOCAL_ONLY = "Local only";
/** A branch that tracks something whose position nothing has fetched yet. */
const NOT_FETCHED = "Not fetched";
const UP_TO_DATE = "Up to date";

export function branchStanding(branch: BranchInfo): BranchStanding {
  const settled = { ahead: 0, behind: 0 };
  // A detached head is not a branch and has no upstream to stand against. The badge names it where
  // a branch name would go, so there is nothing left for the standing to add beside it.
  if (branch.name === null) return { label: null, toneClass: UNSETTLED_TONE_CLASS, ...settled };
  // Tracking nothing and tracking something unfetched are both "no comparison to make", and they
  // are not the same thing to the reader: the first is published nowhere and wants a push, the
  // second wants a fetch before anything can be said about it.
  if (branch.upstream === null) {
    return { label: LOCAL_ONLY, toneClass: UNSETTLED_TONE_CLASS, ...settled };
  }
  switch (branch.sync.state) {
    case "unknown":
      return { label: NOT_FETCHED, toneClass: UNSETTLED_TONE_CLASS, ...settled };
    case "up_to_date":
      return { label: UP_TO_DATE, toneClass: SYNCED_TONE_CLASS, ...settled };
    case "ahead":
      return {
        label: `${branch.sync.ahead} ahead`,
        toneClass: UNSETTLED_TONE_CLASS,
        ahead: branch.sync.ahead,
        behind: 0,
      };
    case "behind":
      return {
        label: `${branch.sync.behind} behind`,
        toneClass: UNSETTLED_TONE_CLASS,
        ahead: 0,
        behind: branch.sync.behind,
      };
    case "diverged":
      return {
        label: `${branch.sync.ahead} ahead, ${branch.sync.behind} behind`,
        toneClass: UNSETTLED_TONE_CLASS,
        ahead: branch.sync.ahead,
        behind: branch.sync.behind,
      };
  }
}
